### DISCUSSION TITLE
Next release date?

### DISCUSSION THREAD

**Discussion by al42and** [DTolm/VkFFT#183] [General]
Hi! Do you have a timeline for the next VkFFT release? Would like to see how it matches up with GROMACS release timings to decide which version to bundle. 1377057ee90eba0dd8a55594beb3cd7098707293 would be nice to have :)

**Comment by DTolm** [comment]
Hello, I think I will merge the current state of the develop branch (with some changes to the test suite and not the library) this month as all the other features require more time to implement. Have you tested this optimization to be useful for GROMACS? It only affects big systems and I am not sure if it has much impact on MI300 (due to it having L3 cache).

**Comment by al42and** [comment]
> I think I will merge the current state of the develop branch (with some changes to the test suite and not the library) this month as all the other features require more time to implement. 

Great!

>  Have you tested this optimization to be useful for GROMACS? It only affects big systems and I am not sure if it has much impact on MI300 (due to it having L3 cache).

A quick benchmark on MI250X has shown speed-up up to ~7% in FFT time (primarily for small systems :thinking:; will need to looks more into that). Don't have MI300 to try.

**Comment by DTolm** [comment]
The improvement may also be due to another commit: https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868 where radix kernels were slightly optimized. The padding mentioned in the AMD commit by default is enabled for systems with all dimension sizes multiplied being more than 2097152.

**Comment by DTolm** [comment]
Hello,

I have implemented a new register assignment logic and a set of optimizations to generated kernels that improved performance quite a bit (especially on Nvidia). I think it may be interesting to run GROMACS with VkFFT on Nvidia hardware to compare these optimizations to cuFFT. Sorry for the delay with the release, I wanted these improvements to be in it.

Best regards,
Dmitrii

**Reply by al42and** [reply]
We sadly don't have a way to cleanly test cuFFT vs. VkFFT (no cuFFT support in SYCL build, no VkFFT support in CUDA build). Not too hard to resolve, but not a priority for us: (

However, the recent changes in `develop` are a bit of a mixed bag for our cases on both CUDA and HIP. Timings below are with kernel serialization enabled; not the most rigorous testing, but seems to be roughly reproducible.

On RTX4070Ti, CUDA 12.4, kernel time (averaged over r2c and c2r and 1e4 steps) in µs:

size (katoms) | FFT size | 066a17c1 | daf09d3 | 1377057 | d0b68bb | 539be29
-- | -- | -- | -- | -- | -- | --
12 | 44x44x44 | 2.05 | 1.89 | 2.05 | 2.08 | 1.81
24 | 96x44x44 | 2.49 | 2.23 | 2.49 | 2.49 | 2.28
48 | 96x96x44 | 3.73 | 3.13 | 3.70 | 3.70 | 4.01
96 | 96x96x96 | 6.11 | 5.23 | 6.14 | 6.14 | 6.72
192 | 192x96x96 | 11.4 | 9.7 | 11.4 | 11.4 | 12.0
384 | 192x192x96 | 27.5 | 23.8 | 27.5 | 27.5 | 27.3
768 | 192x192x192 | 64.2 | 46.4 | 64.3 | 64.3 | 64.3
1536 | 384x192x192 | 214.7 | 189.8 | 214.6 | 214.6 | 216.1

On MI250X, ROCm 6.0.0, kernel time (averaged over r2c and c2r, 1e4 steps, and three runs) in µs:

size (katoms) | FFT size | 066a17c1 | daf09d3 | 1377057 | d0b68bb | 539be29
-- | -- | -- | -- | -- | -- | --
12* | 44x44x44 | 11.1 | 11.3 | 10.5 | 10.4 | 11.7
24* | 96x44x44 | 12.5 | 11.1 | 11.6 | 12.2 | 12.8
48 | 96x96x44 | 14.3 | 12.7 | 12.5 | 12.9 | 14.0
96 | 96x96x96 | 18.3 | 17.0 | 17.2 | 17.4 | 14.4
192 | 192x96x96 | 29.5 | 27.9 | 27.9 | 28.0 | 25.4
384 | 192x192x96 | 53.2 | 50.9 | 54.2 | 54.0 | 49.7
768 | 192x192x192 | 87.7 | 85.0 | 90.4 | 90.1 | 84.0
1536 | 384x192x192 | 176.2 | 174.2 | 184.4 | 184.7 | 173.5

*: 12k and 24k on AMD are quite noisy, so could be skipped.

**Reply by DTolm** [reply]
Hello, thank you for these results! I am still trying to understand the exact systems that are launched because 384x192x192 takes 54MB in R2C/C2R FP32 (384x192x192x4B) and needs to be transferred back and forth three times for each axis, resulting in 54x2x3 = 324MB of transfers. dividing that by 216us or 173us results in bandwidths way out of the hardware specifications (1.5TB/s and 1.9TB/s). Is it possible to generate some ncu reports for the CUDA backend (one step, one run) so I can compare the kernel configurations to cuFFT? This will allow me to fine-tune VkFFT more to the real use case of small multidimensional FFTs and not synthetic benchmarks.

As for the https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868 results, I think it must be an outlier for 4070Ti as there are no changes to kernel configuration specific to this branch. I guess the configuration for these particular kernels didn't change much between 1.3.5 and 1.3.4 (I will double-check the kernels for 48, 96 and 192). On MI250X it does seem to get slightly better overall.

The new kernel configurator is still getting refined, so while some systems may get slightly worse, it is capable of generating all the kernels that were in v1.3.4 so it is just a matter of parameter-tuning.

**Reply by al42and** [reply]
> As for the https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868 results, I think it must be an outlier for 4070Ti as there are no changes to kernel configuration specific to this branch

The benchmarks were run using `for size in $(shuf -e 0012  0024  0048 0096 0192 0384 0768 1536 ); do for f in $(shuf -e  066a17c1 daf09d3  d0b68bb 539be29 1377057)`, so for each size all commits were tested in a different random order. daf09d3 had been the fastest every time, so unlikely just a coincidence. 

>  Is it possible to generate some ncu reports for the CUDA backend (one step, one run) so I can compare the kernel configurations to cuFFT?

539be29f30ef40384f3ccd24d7884166f82cbef8: https://drive.google.com/drive/folders/1y4_T5NfcUfpSd-lexe2oatYJfwpkNI9T?usp=sharing

Kernels timings are slightly larger there, so there could be some measurement/inaccuracies in `nsys`, which I used for the table above.

> This will allow me to fine-tune VkFFT more to the real use case of small multidimensional FFTs and not synthetic benchmarks.

Practically speaking, ~5% difference in FFT is unlikely to influence the overall performance for typical GROMACS use, but it could be rate limiting in some configurations (e.g., when scaling to >4 GPUs without multi-GPU FFT; that would be ~100x100x100 and up). If you want a showcase, I can think of a benchmark setup to highlight the FFT part.

**Reply by DTolm** [reply]
From profiling, I now understood that the kernel times in the table are averaged per axis as well, so if we assume that one FFT takes 3x the kernel time, we get bandwidth of 4070Ti ~500GB/s, which seems correct. Sorry for the confusion.

As for the https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868, there was an incorrect comparison of vendorID that was written as an assignment. So VkFFT used the AMD profile for Nvidia GPUs when deciding how many threads per kernel to use. This resulted in bigger coalescing in strided accesses which turns out to be beneficial in your use case. The mistake was introduced in this update and fixed in the next one. Replacing the contents of vkFFT\vkFFT_PlanManagement\vkFFT_HostFunctions\vkFFT_AxisBlockSplitter.h file with the contents of the attached file should disable the optimization which I think is responsible for the performance drop, so if it actually improves the performance I will rework it for the next release.

I also wanted to ask, are any of the parameters like aimThreads or coalescedMemory changed in GROMACS?
[vkFFT_AxisBlockSplitter.txt](https://github.com/user-attachments/files/17329505/vkFFT_AxisBlockSplitter.txt)

**Reply by al42and** [reply]
> As for the https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868, there was an incorrect comparison of vendorID that was written as an assignment.

Lucky that #166 wasn't merged :)

> Replacing the contents of vkFFT\vkFFT_PlanManagement\vkFFT_HostFunctions\vkFFT_AxisBlockSplitter.h file with the contents of the attached file should disable the optimization which I think is responsible for the performance drop, so if it actually improves the performance I will rework it for the next release.

Looks about right :+1: 

`develop` is 539be29f30ef40384f3ccd24d7884166f82cbef8, `fix` is the same code with your patch added, kernel times (averaged over 3 dims and fwd/bwd):

size (katoms) | develop | fix
-- | -- | --
12 | 2.06 | 2.06
24 | 2.79 | 2.65
48 | 4.87 | 3.81
96 | 8.11 | 6.21
192 | 14.4 | 11.7
384 | 31.3 | 27.6
768 | 72.3 | 53.3
1536 | 223.0 | 191.9

> I also wanted to ask, are any of the parameters like aimThreads or coalescedMemory changed in GROMACS?

`aimThreads` (and not updated in a while :disappointed:): https://gitlab.com/gromacs/gromacs/blob/919c0e1085190bd8cfb4f2077ffcd03ea4337422/src/gromacs/fft/gpu_3dfft_sycl_vkfft.cpp#L167-194

**Reply by DTolm** [reply]
> As for the https://github.com/DTolm/VkFFT/commit/daf09d35cc8b36f33eeda99da2959a470dcc2868, there was an incorrect comparison of vendorID that was written as an assignment.

Well, the assignment bug was caught straight away, but it is still interesting that optimization that was beneficial to synthetic tests was actually detrimental for real applications. You learn something new every day!

> develop is https://github.com/DTolm/VkFFT/commit/539be29f30ef40384f3ccd24d7884166f82cbef8, fix is the same code with your patch added, kernel times (averaged over 3 dims and fwd/bwd):

Nice to see that to get a 25% boost I had to remove code, not add :)

> aimThreads (and not updated in a while 😞):

Is it still better when it is set to 64? I have been optimizing the parameters for 128, so it might be worth checking.

**Reply by al42and** [reply]
> Is it still better when it is set to 64? I have been optimizing the parameters for 128, so it might be worth checking.

At least for serialized kernel on MI250X, ROCm 6.0.0, 64 seems a bit faster than the default of 128 for mid-sized systems, but it is the opposite for smaller systems (same averaging over all dimensions and 10k fwd+bwd roundtrips; three markers == three independent runs on the same GPU):

![Screenshot_20241011_172727](https://github.com/user-attachments/assets/185dbe8c-63e0-4754-9d0e-c7837df51b98)

Would check how it behaves when run in parallel with other kernels later.

**Reply by al42and** [reply]
End-to-end performance, where FFT can overlap on the same GPU with other kernels. More is better.

Different ROCm version from the plot above; also added VkFFT 1.3.1.

Data from 20 runs for 12k (44x44x44) and 48k (96x96x44), and 10 runs for 192k (192x96x96) and 768k (192x192x192).

![grappa_12](https://github.com/user-attachments/assets/2b0c6879-f11e-4189-a112-4743321806a4)
![grappa_48](https://github.com/user-attachments/assets/a79ad42d-117f-494d-a1b8-1b1a5fd95620)
![grappa_192](https://github.com/user-attachments/assets/a8860fa2-99c7-443a-aba6-a40ac557c417)
![grappa_768](https://github.com/user-attachments/assets/efb756ce-b5e9-4b4c-88f0-2da830458cd1)

This is roughly consistent with serialized kernel time measurements (ROCm 6; same data as above, but with 1.3.1 included): 1.3.1 is faster for smaller transforms, 1.3.5 for larger ones:

![Screenshot_20241018_144720](https://github.com/user-attachments/assets/59dcea96-fca4-4ddb-9da2-be6386d1fa68)
